File: src/main/java/org/cip4/xjdf/json/openapi/JsonSchemaConverter.java
package org.cip4.xjdf.json.openapi;

import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.ObjectWriter;
import org.cip4.xjdf.json.openapi.model.Model;
import org.cip4.xjdf.json.openapi.model.Schema;
import org.cip4.xjdf.json.openapi.model.Schemas;
import org.w3c.dom.Document;
import org.w3c.dom.Node;
import org.w3c.dom.NodeList;

import javax.xml.xpath.XPath;
import javax.xml.xpath.XPathConstants;
import javax.xml.xpath.XPathExpressionException;
import java.io.InputStream;
import java.io.OutputStream;
import java.io.OutputStreamWriter;
import java.nio.charset.StandardCharsets;
import java.util.HashMap;
import java.util.Map;

public class JsonSchemaConverter {

    public static final String PREFIX = "#/$defs/";
    private final Document doc;
    private final XPath xPath;

    public static final String XJDF_SCHEMA = "https://schema.cip4.org/jdfschema_2_2/xjdf.json";
    public static final String XJMF_SCHEMA = "https://schema.cip4.org/jdfschema_2_2/xjmf.json";

    public JsonSchemaConverter(InputStream sourceXsd) {
        SchemaReader reader = new SchemaReader();
        this.doc = reader.readXml(sourceXsd);
        this.xPath = reader.getXPath();
    }

    public Schemas convert(OutputStream xjdfOutputStream, OutputStream xjmfOutputStream) {
        ObjectMapper objectMapper = new ObjectMapper();
        objectMapper.setSerializationInclusion(JsonInclude.Include.NON_NULL);
        objectMapper.setSerializationInclusion(JsonInclude.Include.NON_DEFAULT);

        ObjectWriter objectWriter = objectMapper.writerWithDefaultPrettyPrinter();

        Schema xjdfSchema = convertModel("XJDF");
        xjdfSchema.id("https://schema.cip4.org/jdfschema_2_2/xjdf.json");

        try (OutputStreamWriter writer = new OutputStreamWriter(xjdfOutputStream, StandardCharsets.UTF_8)) {
            writer.write(objectWriter.writeValueAsString(xjdfSchema));
        } catch (Exception e) {
            throw new RuntimeException("Error writing XJDF schema", e);
        }

        Map<String, String> xjdfTypes = new HashMap<>();
        xjdfSchema.defs().forEach((key, value) -> xjdfTypes.put(key, xjdfSchema.id() + PREFIX + key));

        Schema xjmfSchema = convertModel("XJMF", xjdfTypes);
        xjmfSchema.id("https://schema.cip4.org/jdfschema_2_2/xjmf.json");

        try (OutputStreamWriter writer = new OutputStreamWriter(xjmfOutputStream, StandardCharsets.UTF_8)) {
            writer.write(objectWriter.writeValueAsString(xjmfSchema));
        } catch (Exception e) {
            throw new RuntimeException("Error writing XJMF schema", e);
        }

        Schemas schemaMap = new Schemas();
        schemaMap.put("XJDF", xjdfSchema);
        schemaMap.put("XJMF", xjmfSchema);

        return schemaMap;
    }

    public Schema convertModel(String root) {
        return convertModel(root, new HashMap<>());
    }

    public Schema convertModel(String root, Map<String, String> referencedTypes) {
        TypeTranslator nameTranslator = new TypeTranslator(PREFIX, referencedTypes);
        Schemas defs = new Schemas();
        convertNamedElement(root, nameTranslator, defs);
        Schema xjdfSchema = defs.remove(root);
        xjdfSchema.schema("https://json-schema.org/draft/2020-12/schema");
        xjdfSchema.defs(defs);
        return xjdfSchema;
    }

    private void convertNamedType(String typeName, TypeTranslator nameTranslator, Schemas schemas) {
        if (schemas.containsKey(typeName) || nameTranslator.isExternalType(typeName)) {
            return;
        }

        try {
            Node typeNode = (Node) xPath.evaluate(
                "/xs:schema/xs:simpleType[@name='" + typeName + "'] | /xs:schema/xs:complexType[@name='" + typeName + "']",
                doc,
                XPathConstants.NODE
            );

            if (typeNode.getLocalName().equals("simpleType")) {
                convertSimpleType(typeNode, nameTranslator, schemas);
            } else {
                convertComplexType(typeNode, nameTranslator, schemas);
            }
        } catch (XPathExpressionException e) {
            throw new RuntimeException("Error evaluating XPath for named type: " + typeName, e);
        }
    }

    private void convertSimpleType(Node typeNode, TypeTranslator nameTranslator, Schemas schemas) {
        Model schema = SimpleType.create(
            typeNode,
            new Context(xPath, nameTranslator, typeNode)
        ).getModel();

        schemas.put(schema.getName(), schema.getSchema());
        convertExtendedTypes(typeNode, nameTranslator, schemas);
    }

    private void convertComplexType(Node typeNode, TypeTranslator nameTranslator, Schemas schemas) {
        Model schema = new ComplexType(new Context(xPath, nameTranslator, typeNode)).getModel();
        schemas.put(schema.getName(), schema.getSchema());

        convertReferencedElements(typeNode, nameTranslator, schemas);
        convertReferencedTypes(typeNode, nameTranslator, schemas);
        convertExtendedTypes(typeNode, nameTranslator, schemas);
    }

    private void convertNamedElement(String elementName, TypeTranslator nameTranslator, Schemas schemas) {
        try {
            Node element = (Node) xPath.evaluate(
                "/xs:schema/xs:element[@name='" + elementName + "']",
                doc,
                XPathConstants.NODE
            );

            convertReferencedTypes(element, nameTranslator, schemas);
            convertSubstituteElements(elementName, nameTranslator, schemas);

            if ((
                element.getAttributes().getNamedItem("abstract") != null
                    && element.getAttributes().getNamedItem("abstract").getNodeValue().equals("true")
            )
                || element.getAttributes().getNamedItem("substitutionGroup") != null
            ) {
                return;
            }

            Model schema = new NamedElement(new Context(xPath, nameTranslator, element)).getModel();
            schemas.put(schema.getName(), schema.getSchema());

            convertReferencedElements(element, nameTranslator, schemas);
            convertExtendedTypes(element, nameTranslator, schemas);
        } catch (XPathExpressionException e) {
            throw new RuntimeException("Error evaluating XPath for named element: " + elementName, e);
        }
    }

    private void convertReferencedTypes(Node node, TypeTranslator nameTranslator, Schemas schemas) {
        try {
            NodeList types = (NodeList) xPath.evaluate(".//@type | .//@itemType", node, XPathConstants.NODESET);

            for (int i = 0; i < types.getLength(); i++) {
                String typeName = types.item(i).getNodeValue();
                convertNamedType(typeName, nameTranslator, schemas);
            }
        } catch (XPathExpressionException e) {
            throw new RuntimeException("Error evaluating XPath for referenced types", e);
        }
    }

    private void convertExtendedTypes(Node element, TypeTranslator nameTranslator, Schemas schemas) {
        try {
            NodeList elements = (NodeList) xPath.evaluate(".//@base", element, XPathConstants.NODESET);

            for (int i = 0; i < elements.getLength(); i++) {
                convertNamedType(elements.item(i).getNodeValue(), nameTranslator, schemas);
            }
        } catch (XPathExpressionException e) {
            throw new RuntimeException("Error evaluating XPath for extended types", e);
        }
    }

    private void convertReferencedElements(Node node, TypeTranslator nameTranslator, Schemas schemas) {
        try {
            NodeList elements = (NodeList) xPath.evaluate(".//xs:element/@ref", node, XPathConstants.NODESET);

            for (int i = 0; i < elements.getLength(); i++) {
                String elementName = elements.item(i).getNodeValue();
                if (!schemas.containsKey(elementName) && !nameTranslator.isExternalType(elementName)) {
                    convertNamedElement(elementName, nameTranslator, schemas);
                }
            }
        } catch (XPathExpressionException e) {
            throw new RuntimeException("Error evaluating XPath for referenced elements", e);
        }
    }

    private void convertSubstituteElements(String elementName, TypeTranslator nameTranslator, Schemas schemas) {
        try {
            NodeList elements = (NodeList) xPath.evaluate(
                "/xs:schema/xs:element[@substitutionGroup='" + elementName + "']/@name",
                doc,
                XPathConstants.NODESET
            );

            for (int i = 0; i < elements.getLength(); i++) {
                convertNamedElement(elements.item(i).getNodeValue(), nameTranslator, schemas);
            }
        } catch (XPathExpressionException e) {
            throw new RuntimeException("Error evaluating XPath for substitute elements", e);
        }
    }
}

File: src/main/kotlin/org/cip4/xjdf/openapi/JsonSchemaConverter.kt
/*
 * The CIP4 Software License, Version 1.0
 *
 * Copyright (c) 2001-2021 The International Cooperation for the Integration of
 * Processes in Prepress, Press and Postpress (CIP4). All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without modification,
 * are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this
 * list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 * this list of conditions and the following disclaimer in the documentation and/or
 * other materials provided with the distribution.
 *
 * 3. The end-user documentation included with the redistribution, if any, must
 * include the following acknowledgment: &quot;This product includes software developed
 * by the The International Cooperation for the Integration of Processes in
 * Prepress, Press and Postpress (www.cip4.org)&quot; Alternately, this acknowledgment
 * may appear in the software itself, if and wherever such third-party
 * acknowledgments normally appear.
 *
 * 4. The names &quot;CIP4&quot; and &quot;The International Cooperation for the Integration of
 * Processes in Prepress, Press and Postpress&quot; must not be used to endorse or
 * promote products derived from this software without prior written permission.
 * For written permission, please contact info@cip4.org
 *
 * 5. Products derived from this software may not be called &quot;CIP4&quot;, nor may &quot;CIP4&quot;
 * appear in their name, without prior written permission of the CIP4 organization
 *
 * &quot;CIP4&quot; and &quot;The International Cooperation for the Integration of Processes in
 * Prepress, Press and Postpress&quot; are trademarks of The International Cooperation
 * for the Integration of Processes in Prepress, Press and Postpress (CIP4).  All
 * other names and brands are the property of their respective owners.
 *
 * THIS SOFTWARE IS PROVIDED ``AS IS'' AND ANY EXPRESSED OR IMPLIED WARRANTIES,
 * INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
 * FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. CIP4 AND ITS CONTRIBUTORS MAKE
 * NO REPRESENTATION THAT THE USE OF THIS SOFTWARE OR DATA WILL NOT INFRINGE ANY
 * PATENTS, COPYRIGHTS, TRADEMARKS, OR OTHER RIGHTS. IN NO EVENT SHALL CIP4 OR ITS
 * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY,
 * OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
 * IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY
 * OF SUCH DAMAGE.
 * ====================================================================
 *
 * This software consists of voluntary contributions made by many individuals on
 * behalf of the The International Cooperation for the Integration of Processes in
 * Prepress, Press and Postpress and was originally based on software some of which
 * contributed by and copyrighted by Adobe Systems and Man Roland. copyright (c)
 * 2000-2001, Adobe Systems copyright (c) 2001-2001, Man Roland
 *
 * For more information on The International Cooperation for the Integration of
 * Processes in Prepress, Press and Postpress , please see &lt;http://www.cip4.org/&gt;.
 */

package org.cip4.xjdf.openapi

import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import org.cip4.xjdf.openapi.model.Schema
import org.cip4.xjdf.openapi.model.Schemas
import org.w3c.dom.Document
import org.w3c.dom.Node
import org.w3c.dom.NodeList
import org.xml.sax.InputSource
import java.io.InputStream
import java.io.OutputStream
import javax.xml.parsers.DocumentBuilderFactory
import javax.xml.xpath.XPath
import javax.xml.xpath.XPathConstants
import javax.xml.xpath.XPathFactory

class JsonSchemaConverter(sourceXsd: InputStream) {
    private var doc: Document
    private val xPath: XPath

    init {
        doc = readXml(sourceXsd)
        val xpFactory = XPathFactory.newInstance()
        xPath = xpFactory.newXPath()
        xPath.namespaceContext = XsdNamespaceContext()
    }

    private fun readXml(sourceXsd: InputStream): Document {
        val dbFactory = DocumentBuilderFactory.newInstance()
        dbFactory.isNamespaceAware = true
        val dBuilder = dbFactory.newDocumentBuilder()
        val xmlInput = InputSource(sourceXsd)

        return dBuilder.parse(xmlInput)
    }

    fun convert(
        xjdfOutputStream: OutputStream,
        xjmfOutputStream: OutputStream
    ) : TypeMap {
        val format = Json {
            encodeDefaults = false
            prettyPrint = true
        }

        val xjdfSchema = convertModel("XJDF")
        xjdfSchema.`$id` = "https://schema.cip4.org/jdfschema_2_2/xjdf.json"
        xjdfOutputStream.writer().use { writer ->
            writer.write(format.encodeToString(xjdfSchema))
        }

        val xjdfTypes = xjdfSchema.`$defs`.keys.associateWith {
            // "xjdf-schema.yml#/\$defs/$it"
            xjdfSchema.`$id` + "#/\$defs/" + it
        }

        val xjmfSchema = convertModel("XJMF", xjdfTypes)
        xjmfSchema.`$id` = "https://schema.cip4.org/jdfschema_2_2/xjmf.json"
        xjmfOutputStream.writer().use { writer ->
            writer.write(format.encodeToString(xjmfSchema))
        }

        return TypeMap(mapOf(
            Pair("XJDF", xjdfSchema),
            Pair("XJMF", xjmfSchema)
        ))
    }

    fun convertModel(root: String, referencedTypes: Map<String, String> = mapOf()): Schema {
        val nameTranslator = TypeTranslator("#/\$defs/", referencedTypes)
        val defs = Schemas()
        convertNamedElement(root, nameTranslator, defs)
        val xjdfSchema = defs[root]!!
        defs.remove(root)
        xjdfSchema.`$schema` = "https://json-schema.org/draft/2020-12/schema"
        xjdfSchema.`$defs` = defs
        return xjdfSchema
    }

    private fun convertReferencedTypes(node: Node, nameTranslator: TypeTranslator, schemas: Schemas) {
        val types = xPath.evaluate(
            ".//@type | .//@itemType",
            node,
            XPathConstants.NODESET
        ) as NodeList

        for (i in 0 until types.length) {
            val typeName = types.item(i).nodeValue
            convertNamedType(typeName, nameTranslator, schemas)
        }
    }

    private fun convertNamedType(typeName: String, nameTranslator: TypeTranslator, schemas: Schemas) {
        if (schemas.containsKey(typeName)
            || nameTranslator.isExternalType(typeName)
        ) {
            return
        }
        val typeNode = xPath.evaluate(
            "/xs:schema/xs:simpleType[@name='$typeName'] | /xs:schema/xs:complexType[@name='$typeName']",
            doc,
            XPathConstants.NODE
        ) as Node

        if (typeNode.localName == "simpleType") {
            convertSimpleType(typeNode, nameTranslator, schemas)
        } else {
            convertComplexType(typeNode, nameTranslator, schemas)
        }
    }

    private fun convertSimpleType(typeNode: Node, nameTranslator: TypeTranslator, schemas: Schemas) {
        val schema = SimpleType.Factory.create(
            typeNode,
            Context(xPath, nameTranslator, typeNode)
        ).getModel()
        schemas[schema.name] = schema.schema

        convertExtendedTypes(typeNode, nameTranslator, schemas)
    }

    private fun convertComplexType(typeNode: Node, nameTranslator: TypeTranslator, schemas: Schemas) {
        val schema = ComplexType(Context(xPath, nameTranslator, typeNode)).getModel()
        schemas[schema.name] = schema.schema

        convertReferencedElements(typeNode, nameTranslator, schemas)
        convertReferencedTypes(typeNode, nameTranslator, schemas)
        convertExtendedTypes(typeNode, nameTranslator, schemas)
    }

    private fun convertNamedElement(
        elementName: String,
        nameTranslator: TypeTranslator,
        schemas: Schemas
    ) {
        val element = xPath.evaluate(
            "/xs:schema/xs:element[@name='$elementName']",
            doc,
            XPathConstants.NODE
        ) as Node

        convertReferencedTypes(element, nameTranslator, schemas)

        convertSubstituteElements(elementName, nameTranslator, schemas)

        if (element.attributes.getNamedItem("abstract")?.nodeValue == "true"
            || element.attributes.getNamedItem("substitutionGroup") != null
        ) {
            return
        }

        NamedElement(Context(xPath, nameTranslator, element)).getModel().let {
            schemas[it.name] = it.schema
        }
        convertReferencedElements(element, nameTranslator, schemas)

        convertExtendedTypes(element, nameTranslator, schemas)
    }

    private fun convertExtendedTypes(
        element: Node,
        nameTranslator: TypeTranslator,
        schemas: Schemas
    ) {
        val elements = xPath.evaluate(
            ".//@base",
            element,
            XPathConstants.NODESET
        ) as NodeList

        for (i in 0 until elements.length) {
            convertNamedType(elements.item(i).nodeValue!!, nameTranslator, schemas)
        }
    }

    private fun convertReferencedElements(node: Node, nameTranslator: TypeTranslator, schemas: Schemas) {
        val elements = xPath.evaluate(
            ".//xs:element/@ref",
            node,
            XPathConstants.NODESET
        ) as NodeList

        for (i in 0 until elements.length) {
            val elementName = elements.item(i).nodeValue
            if (!schemas.containsKey(elementName)
                && !nameTranslator.isExternalType(elementName)
            ) {
                convertNamedElement(elementName, nameTranslator, schemas)
            }
        }
    }

    private fun convertSubstituteElements(
        elementName: String,
        nameTranslator: TypeTranslator,
        schemas: Schemas
    ): Boolean {
        val elements = xPath.evaluate(
            "/xs:schema/xs:element[@substitutionGroup='$elementName']/@name",
            doc,
            XPathConstants.NODESET
        ) as NodeList

        for (i in 0 until elements.length) {
            convertNamedElement(elements.item(i).nodeValue, nameTranslator, schemas)
        }

        return elements.length > 0
    }

}